dynamic_programming/fibonacci_numbers.nim

```diff
@@ -0,0 +1,152 @@
+## Fibonacci Numbers
+## ===========================================================================
+## Fibonacci numbers are numbers that are part of the Fibonacci sequence.
+## The Fibonacci sequence generally starts with 0 and 1, every next number
+## is the sum of the two preceding numbers:
+##   F(0) = 0;
+##   F(1) = 1;
+##   F(n) = F(n-1) + F(n-2);
+##
+##  References:
+##    https://en.wikipedia.org/wiki/Fibonacci_sequence
+##    https://oeis.org/A000045
+{.push raises: [].}
+
+runnableExamples:
+
+  doAssert fibonacciSeqIterative(5.Natural) == @[Natural(0), 1, 1, 2, 3]
+
+  doAssert fibonacciClosure(4.Natural) == Natural(3)
+
+  for f in fibonacciIterator(1.Natural..4.Natural):
+    echo f # 1, 1, 2, 3
+
+import std/math
+
+func fibonacciRecursive*(nth: Natural): Natural =
+  ## Calculates the n-th fibonacci number in a recursive manner.
+  ## Recursive algorithm is extremely slow for bigger numbers of n.
+  if nth <= 1: return nth
+  fibonacciRecursive(nth-2) + fibonacciRecursive(nth-1)
+
+func makeFibClosure(): proc(): Natural =
+  ## Closure constructor.
+  ## Returns procedure which can be called to get the next fibonacci number.
+  var prev = 0
+  var current = 1
+  proc(): Natural =
+    swap(prev, current)
+    current += prev
+    result = current
+
+proc fibonacciClosure*(nth: Natural): Natural =
+  ## Calculates the n-th fibonacci number with use of a closure.
+  if nth <= 1: return nth
+  let fib = makeFibClosure()
+  for _ in 2..<nth:
+    discard fib()
+  fib()
+
+proc fibonacciSeqClosure*(n: Natural): seq[Natural] =
+  ## Generates a list of n first fibonacci numbers with use of a closure.
+  result = newSeq[Natural](n)
+  result[0] = 0
+  if n > 1:
+    result[1] = 1
+    let fib = makeFibClosure()
+    for i in 2..<n:
+      result[i] = fib()
+
+func fibonacciSeqIterative*(n: Natural): seq[Natural] =
+  ## Generates a list of n first fibonacci numbers in iterative manner.
+  result = newSeq[Natural](n)
+  result[0] = 0
+  if n > 1:
+    result[1] = 1
+    for i in 2..<n:
+      result[i] = result[i-1] + result[i-2]
+
+iterator fibonacciIterator*(s: HSlice[Natural, Natural]): Natural =
+  ## Nim iterator.
+  ## Returns fibonacci numbers from F(s.a) to F(s.b).
+  var prev = 0
+  var current = 1
+  for i in 0..s.b:
+    if i <= 1:
+      if i >= s.a: yield i
+    else:
+      swap(prev, current)
+      current += prev
+      if i >= s.a: yield current
+
+func fibonacciBinet*(nth: Natural): Natural =
+  ## Calculates the n-th fibonacci number in constant time O(1) with use of Binet's formula.
```

Comment:
*Approximates. Binet's formula has only been rediscovered by Binet (It was previously known by Euler and even maybe by De Moivre). I would remove this nomination and add a reference to the formula in the docstring.

Reply:
- [x] Ok. Besides, it's not really an algorithm in the traditional sense. I'll remove it.

Reply:
I would keep the algorithm, it it is educational to compare the Fibonacci numbers with the approximation. I would just change the name. The description is misleading because we can not compute Fibonacci numbers exactly in O(1). The formula requires infinite precision to output the exact value.

Reply:
> Binet's formula has only been rediscovered by Binet...

I used name of Binet here, because wikipedia says:
> the Fibonacci numbers have a closed-form expression. It has become known as Binet's formula.

What names should I use instead?
Currently, the procedure is `fibonacciBinet` and in the tests I refer to it as "Constant time formula".
Are they both should be changed?

Reply:
The French wikipedia page explains that it has been rediscovered by Binet.

I would give them a descriptive name: `fibonacciClosedFormApproximation`